src/Workspaces/Core/Portable/CodeActions/Operations/CodeActionOperation.cs

```diff
@@ -25,10 +27,16 @@ public virtual void Apply(Workspace workspace, CancellationToken cancellationTok
         {
         }
 
-        internal virtual bool TryApply(Workspace workspace, IProgressTracker progressTracker, CancellationToken cancellationToken)
+        /// <summary>
+        /// Called by the host environment to apply the effect of the operation.
+        /// This method is guaranteed to be called on the UI thread.
+        /// </summary>
+        internal virtual Task<bool> TryApplyAsync(Workspace workspace, IProgressTracker progressTracker, CancellationToken cancellationToken)
         {
+            // It is a requirement that this method be alled on the UI thread.  So it's safe for us to call
+            // into .Apply without any threading operations here.
             this.Apply(workspace, cancellationToken);
-            return true;
+            return SpecializedTasks.True;
```

Comment:
@sharwell i dont' know what we can do better here given that this is at the Workspace layer.